Implement areInlineCompatible for SystemZ using feature bitset

chavandres opened the pull request:
## What?
Implement `areInlineCompatible` for the SystemZ target using FeatureBitset comparison.

## Why?
The default implementation in `TargetTransformInfoImpl.h` makes a string comparison and only inlines when the target-cpu and the target-features for caller and callee are the same. We are missing out on optimizations when the callee has a subset of features of the caller.

## How?
Get the FeatureBitset of the caller and callee and check when callee is a subset or equal to the caller's features. It's a similar implementation to ARM, PowerPC...

## Testing?
Test cases check for when the callee is a subset of the caller, when it's not a subset and when both are equals.

chavandres committed:
[SystemZ] Implement areInlineCompatible for SystemZ TTI

uweigand commented on llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:
```diff
@@ -422,6 +422,20 @@ bool SystemZTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
+
+  // Check that target features from the callee are subset or
+  // equal to the caller's features.
+  return (CalleeBits == CallerBits) || (CalleeBits < CallerBits);
```
As far as I can tell, '<' on the `FeatureBitset` type does *not* implement a subset test.  To test for that, you'd need something like
```
CalleeBits & CallerBits == CalleeBits
```

However, in fact allowing any subsets is semantically incorrect.  In particular, the `vector` feature changes the ABI, so we need to be careful here - we may need something similar to the AVX-512 tests in X86.  (Or, maybe, we can at least support inlining as long as both caller and callee agree on the vector feature.)

There might be other features that could be problematic as well, I need to review in more detail.

chavandres replied:
Got it. I agree with you on the danger of inlining without agreeing on the `vector` feature. 

So the implementation can change to only inline when the bitset are the same, unless we add logic to agree on the `vector` feature assuming that is the only problematic case?

uweigand replied:
Inlining when the bitsets are the same would certainly be correct, and we might want to do that as an interim solution.  But is that actually any different from the current (default) behavior?

chavandres replied:
In terms of behavior it will be the same as the default implementation. But given that the default uses string comparion, checking bitset equality is a cleaner solution.

uweigand replied:
Sure.  I guess that would be fine with me, thanks.

chavandres replied:
Thank you. I updated the code and test case to only inline when both callee and caller have equal bitset.

uweigand commented:
Can you please fix the formatting issues clang-format has complained about?   Also, maybe expand the comment to say that we're currently only supporting identical feature sets, but that restriction should be relaxed in the future. Otherwise, this looks good.

chavandres commented:
Fixed the formatting issues, plus expanded the comment that we might support subsets in the future. I rebased the branch with the latest main and squashed my changes to single commit.

uweigand reviewed (approved):
LGTM now, thanks!
